src/Sylius/Bundle/CoreBundle/OAuth/UserProvider.php

```diff
@@ -132,21 +141,34 @@ private function createUserByOAuthUserResponse(UserResponseInterface $response)
     {
         /** @var SyliusUserInterface $user */
         $user = $this->userFactory->createNew();
-        /** @var CustomerInterface $customer */
-        $customer = $this->customerFactory->createNew();
+
+        $customer = $this->customerRepository->findOneBy(['emailCanonical' => $response->getEmail()]);
+
+        if (null === $customer) {
+            /** @var CustomerInterface $customer */
+            $customer = $this->customerFactory->createNew();
+        }
+
         $user->setCustomer($customer);
 
         // set default values taken from OAuth sign-in provider account
         if (null !== $email = $response->getEmail()) {
             $customer->setEmail($email);
         }
 
-        if (null !== $realName = $response->getRealName()) {
-            $customer->setFirstName($realName);
+        if (null !== $name = $response->getFirstName()) {
+            $customer->setFirstName($name);
+        } else {
+            if (null !== $realName = $response->getRealName()) {
+                $customer->setFirstName($realName);
+            }
         }
-
-        if (!$user->getUsername()) {
-            $user->setUsername($response->getEmail() ?: $response->getNickname());
+        if (null !== $lastName = $response->getLastName()) {
+            $customer->setLastName($lastName);
+        } else {
```

Comment:
This `else` looks a bit wrong, cause you are not setting in it `lastName` but the `username`.